File: tests/cpp/gx_character_plane_test.cpp
#include "common/endian.h"
#include "machine/bus/io.h"
#include "machine/devices/gx/character_plane.h"
#include "machine/memory/memory.h"
#include "render/backend/backend.h"
#include "render/backend/pass/library.h"
#include "render/backend/software/gx_character_plane.h"
#include "render/gx/character_plane_resources.h"

#include <array>
#include <cstdint>
#include <stdexcept>

namespace {

void require(bool condition, const char* message) {
	if (!condition) {
		throw std::runtime_error(message);
	}
}

struct CharacterPlaneHarness {
	std::array<uint8_t, 1> emptyRom{{0}};
	bmsx::Memory memory;
	bmsx::GxCharacterPlane plane;

	CharacterPlaneHarness()
		: memory(bmsx::MemoryInit{{emptyRom.data(), 0u}, {emptyRom.data(), 0u}})
		, plane(memory) {
		plane.reset();
	}
};

void testRawIndexedRegisterfiles() {
	CharacterPlaneHarness harness;
	const bmsx::GxCharacterPlaneOutput& resetOutput = harness.plane.readDeviceOutput();

	require(resetOutput.controlWord == 0u, "GX character plane resets disabled");
	require(bmsx::readLE32(resetOutput.paletteBytes.data()) == 0u, "GX character palette SRAM resets clear");
	require(bmsx::readLE32(resetOutput.glyphBytes.data()) == 0u, "GX character glyph SRAM resets clear");
	require(bmsx::readLE32(resetOutput.cellBytes.data()) == 0u, "GX character cell SRAM resets clear");

	harness.memory.writeMappedU32LE(bmsx::IO_GX_CHARACTER_CONTROL, 0x80000000u | bmsx::GX_CHARACTER_PLANE_CONTROL_ENABLE);
	require(harness.memory.readMappedU32LE(bmsx::IO_GX_CHARACTER_CONTROL) == 0x80000001u, "GX character control retains raw word");

	const bmsx::u32 paletteRevisionBeforeWrite = harness.plane.readDeviceOutput().paletteRevision;
	const bmsx::u32 glyphRevisionBeforePaletteWrite = harness.plane.readDeviceOutput().glyphRevision;
	const bmsx::u32 cellRevisionBeforePaletteWrite = harness.plane.readDeviceOutput().cellRevision;
	harness.memory.writeMappedU32LE(bmsx::IO_GX_CHARACTER_PALETTE_ADDRESS, static_cast<bmsx::u32>(bmsx::GX_CHARACTER_PLANE_PALETTE_WORDS - 1u));
	harness.memory.writeMappedU32LE(bmsx::IO_GX_CHARACTER_PALETTE_DATA, 0x89abcdefu);
	harness.memory.writeMappedU32LE(bmsx::IO_GX_CHARACTER_PALETTE_DATA, 0x01234567u);
	require(harness.memory.readMappedU32LE(bmsx::IO_GX_CHARACTER_PALETTE_ADDRESS) == 1u, "GX character palette address wraps");
	require(bmsx::readLE32(resetOutput.paletteBytes.data() + (bmsx::GX_CHARACTER_PLANE_PALETTE_WORDS - 1u) * bmsx::GX_CHARACTER_PLANE_WORD_BYTES) == 0x89abcdefu, "GX character palette stores high raw word");
	require(bmsx::readLE32(resetOutput.paletteBytes.data()) == 0x01234567u, "GX character palette stores wrapped raw word");
	require(harness.plane.readDeviceOutput().paletteRevision != paletteRevisionBeforeWrite, "GX character palette write publishes palette revision");
	require(harness.plane.readDeviceOutput().glyphRevision == glyphRevisionBeforePaletteWrite, "GX character palette write preserves glyph revision");
	require(harness.plane.readDeviceOutput().cellRevision == cellRevisionBeforePaletteWrite, "GX character palette write preserves cell revision");

	const bmsx::u32 paletteRevisionBeforeGlyphWrite = harness.plane.readDeviceOutput().paletteRevision;
	const bmsx::u32 glyphRevisionBeforeWrite = harness.plane.readDeviceOutput().glyphRevision;
	const bmsx::u32 cellRevisionBeforeGlyphWrite = harness.plane.readDeviceOutput().cellRevision;
	harness.memory.writeMappedU32LE(bmsx::IO_GX_CHARACTER_GLYPH_ADDRESS, static_cast<bmsx::u32>(bmsx::GX_CHARACTER_PLANE_GLYPH_WORDS - 1u));
	harness.memory.writeMappedU32LE(bmsx::IO_GX_CHARACTER_GLYPH_DATA, 0xfedcba98u);
	harness.memory.writeMappedU32LE(bmsx::IO_GX_CHARACTER_GLYPH_DATA, 0x76543210u);
	require(harness.memory.readMappedU32LE(bmsx::IO_GX_CHARACTER_GLYPH_ADDRESS) == 1u, "GX character glyph address wraps");
	require(bmsx::readLE32(resetOutput.glyphBytes.data() + (bmsx::GX_CHARACTER_PLANE_GLYPH_WORDS - 1u) * bmsx::GX_CHARACTER_PLANE_WORD_BYTES) == 0xfedcba98u, "GX character glyph stores high raw word");
	require(bmsx::readLE32(resetOutput.glyphBytes.data()) == 0x76543210u, "GX character glyph stores wrapped raw word");
	require(harness.plane.readDeviceOutput().paletteRevision == paletteRevisionBeforeGlyphWrite, "GX character glyph write preserves palette revision");
	require(harness.plane.readDeviceOutput().glyphRevision != glyphRevisionBeforeWrite, "GX character glyph write publishes glyph revision");
	require(harness.plane.readDeviceOutput().cellRevision == cellRevisionBeforeGlyphWrite, "GX character glyph write preserves cell revision");

	const bmsx::u32 paletteRevisionBeforeCellWrite = harness.plane.readDeviceOutput().paletteRevision;
	const bmsx::u32 glyphRevisionBeforeCellWrite = harness.plane.readDeviceOutput().glyphRevision;
	const bmsx::u32 cellRevisionBeforeWrite = harness.plane.readDeviceOutput().cellRevision;
	harness.memory.writeMappedU32LE(bmsx::IO_GX_CHARACTER_CELL_ADDRESS, static_cast<bmsx::u32>(bmsx::GX_CHARACTER_PLANE_CELL_WORDS - 1u));
	harness.memory.writeMappedU32LE(bmsx::IO_GX_CHARACTER_CELL_DATA, 0xdeadbeefu);
	harness.memory.writeMappedU32LE(bmsx::IO_GX_CHARACTER_CELL_DATA, 0x13579bdfu);
	require(harness.memory.readMappedU32LE(bmsx::IO_GX_CHARACTER_CELL_ADDRESS) == 1u, "GX character cell address wraps");
	require(bmsx::readLE32(resetOutput.cellBytes.data() + (bmsx::GX_CHARACTER_PLANE_CELL_WORDS - 1u) * bmsx::GX_CHARACTER_PLANE_WORD_BYTES) == 0xdeadbeefu, "GX character cell stores high raw word");
	require(bmsx::readLE32(resetOutput.cellBytes.data()) == 0x13579bdfu, "GX character cell stores wrapped raw word");
	require(harness.plane.readDeviceOutput().paletteRevision == paletteRevisionBeforeCellWrite, "GX character cell write preserves palette revision");
	require(harness.plane.readDeviceOutput().glyphRevision == glyphRevisionBeforeCellWrite, "GX character cell write preserves glyph revision");
	require(harness.plane.readDeviceOutput().cellRevision != cellRevisionBeforeWrite, "GX character cell write publishes cell revision");

	const bmsx::GxCharacterPlaneOutput& output = harness.plane.readDeviceOutput();
	require(&output == &resetOutput, "GX character output is retained");
	const bmsx::u32 paletteRevisionBeforeReset = output.paletteRevision;
	const bmsx::u32 glyphRevisionBeforeReset = output.glyphRevision;
	const bmsx::u32 cellRevisionBeforeReset = output.cellRevision;
	harness.plane.reset();
	require(harness.plane.readDeviceOutput().controlWord == 0u, "GX character reset disables plane after writes");
	require(bmsx::readLE32(output.paletteBytes.data()) == 0u, "GX character reset clears written palette SRAM");
	require(bmsx::readLE32(output.glyphBytes.data()) == 0u, "GX character reset clears written glyph SRAM");
	require(bmsx::readLE32(output.cellBytes.data()) == 0u, "GX character reset clears written cell SRAM");
	require(output.paletteRevision != paletteRevisionBeforeReset, "GX character reset republishes palette SRAM");
	require(output.glyphRevision != glyphRevisionBeforeReset, "GX character reset republishes glyph SRAM");
	require(output.cellRevision != cellRevisionBeforeReset, "GX character reset republishes cell SRAM");
}

void testDataReadNormalization() {
	CharacterPlaneHarness harness;

	harness.memory.writeMappedU32LE(bmsx::IO_GX_CHARACTER_PALETTE_ADDRESS, 0xffffffffu);
	harness.memory.writeMappedU32LE(bmsx::IO_GX_CHARACTER_PALETTE_DATA, 0xa1b2c3d4u);
	harness.memory.writeMappedU32LE(bmsx::IO_GX_CHARACTER_PALETTE_ADDRESS, 0xffffffffu);
	require(harness.memory.readMappedU32LE(bmsx::IO_GX_CHARACTER_PALETTE_DATA) == 0xa1b2c3d4u, "GX character palette data read uses low address bits");
	require(harness.memory.readMappedU32LE(bmsx::IO_GX_CHARACTER_PALETTE_ADDRESS) == 0u, "GX character palette data read advances wrapped address");

	harness.memory.writeMappedU32LE(bmsx::IO_GX_CHARACTER_CELL_ADDRESS, 0xffffffffu);
	harness.memory.writeMappedU32LE(bmsx::IO_GX_CHARACTER_CELL_DATA, 0x2468ace0u);
	harness.memory.writeMappedU32LE(bmsx::IO_GX_CHARACTER_CELL_ADDRESS, 0xffffffffu);
	require(harness.memory.readMappedU32LE(bmsx::IO_GX_CHARACTER_CELL_DATA) == 0x2468ace0u, "GX character cell data read uses modulo address");
	require(harness.memory.readMappedU32LE(bmsx::IO_GX_CHARACTER_CELL_ADDRESS) == (0xffffffffu % bmsx::GX_CHARACTER_PLANE_CELL_WORDS) + 1u, "GX character cell data read advances modulo address");
}

void testStateRestore() {
	CharacterPlaneHarness harness;

	harness.memory.writeMappedU32LE(bmsx::IO_GX_CHARACTER_CONTROL, 0xf0000001u);
	harness.memory.writeMappedU32LE(bmsx::IO_GX_CHARACTER_PALETTE_ADDRESS, 3u);
	harness.memory.writeMappedU32LE(bmsx::IO_GX_CHARACTER_PALETTE_DATA, 0x80007c00u);
	harness.memory.writeMappedU32LE(bmsx::IO_GX_CHARACTER_GLYPH_ADDRESS, 65u);
	harness.memory.writeMappedU32LE(bmsx::IO_GX_CHARACTER_GLYPH_DATA, 0x00f99f90u);
	harness.memory.writeMappedU32LE(bmsx::IO_GX_CHARACTER_CELL_ADDRESS, 321u);
	harness.memory.writeMappedU32LE(bmsx::IO_GX_CHARACTER_CELL_DATA, 0xabcd1241u);
	const bmsx::GxCharacterPlaneState state = harness.plane.captureState();
	const bmsx::GxCharacterPlaneOutput& output = harness.plane.readDeviceOutput();
	const auto* paletteBytes = &output.paletteBytes;
	const auto* glyphBytes = &output.glyphBytes;
	const auto* cellBytes = &output.cellBytes;

	harness.memory.writeMappedU32LE(bmsx::IO_GX_CHARACTER_CONTROL, 0u);
	harness.memory.writeMappedU32LE(bmsx::IO_GX_CHARACTER_PALETTE_ADDRESS, 3u);
	harness.memory.writeMappedU32LE(bmsx::IO_GX_CHARACTER_PALETTE_DATA, 0u);
	harness.memory.writeMappedU32LE(bmsx::IO_GX_CHARACTER_GLYPH_ADDRESS, 65u);
	harness.memory.writeMappedU32LE(bmsx::IO_GX_CHARACTER_GLYPH_DATA, 0u);
	harness.memory.writeMappedU32LE(bmsx::IO_GX_CHARACTER_CELL_ADDRESS, 321u);
	harness.memory.writeMappedU32LE(bmsx::IO_GX_CHARACTER_CELL_DATA, 0u);
	const bmsx::u32 changedRevision = harness.plane.readDeviceOutput().cellRevision;

	harness.plane.restoreState(state);
	const bmsx::GxCharacterPlaneOutput& restored = harness.plane.readDeviceOutput();
	require(&restored.paletteBytes == paletteBytes, "GX character palette output storage remains stable");
	require(&restored.glyphBytes == glyphBytes, "GX character glyph output storage remains stable");
	require(&restored.cellBytes == cellBytes, "GX character cell output storage remains stable");
	require(restored.controlWord == 0xf0000001u, "GX character control restores");
	require(harness.memory.readMappedU32LE(bmsx::IO_GX_CHARACTER_PALETTE_ADDRESS) == 4u, "GX character palette address restores");
	require(harness.memory.readMappedU32LE(bmsx::IO_GX_CHARACTER_GLYPH_ADDRESS) == 66u, "GX character glyph address restores");
	require(harness.memory.readMappedU32LE(bmsx::IO_GX_CHARACTER_CELL_ADDRESS) == 322u, "GX character cell address restores");
	require(bmsx::readLE32(restored.paletteBytes.data() + 3u * bmsx::GX_CHARACTER_PLANE_WORD_BYTES) == 0x80007c00u, "GX character palette SRAM restores");
	require(bmsx::readLE32(restored.glyphBytes.data() + 65u * bmsx::GX_CHARACTER_PLANE_WORD_BYTES) == 0x00f99f90u, "GX character glyph SRAM restores");
	require(bmsx::readLE32(restored.cellBytes.data() + 321u * bmsx::GX_CHARACTER_PLANE_WORD_BYTES) == 0xabcd1241u, "GX character cell SRAM restores");
	require(restored.cellRevision != changedRevision, "GX character restore republishes cell SRAM");
}

void testBackendResourcesAndSoftwareComposition() {
	CharacterPlaneHarness harness;
	harness.memory.writeMappedU32LE(bmsx::IO_GX_CHARACTER_PALETTE_ADDRESS, 1u);
	harness.memory.writeMappedU32LE(bmsx::IO_GX_CHARACTER_PALETTE_DATA, 0x801fu);
	harness.memory.writeMappedU32LE(bmsx::IO_GX_CHARACTER_PALETTE_DATA, 0xfc00u);
	harness.memory.writeMappedU32LE(bmsx::IO_GX_CHARACTER_GLYPH_ADDRESS, 65u);
	harness.memory.writeMappedU32LE(bmsx::IO_GX_CHARACTER_GLYPH_DATA, 1u);
	harness.memory.writeMappedU32LE(bmsx::IO_GX_CHARACTER_CELL_ADDRESS, 0u);
	harness.memory.writeMappedU32LE(bmsx::IO_GX_CHARACTER_CELL_DATA, 0x2141u);
	harness.memory.writeMappedU32LE(bmsx::IO_GX_CHARACTER_CELL_ADDRESS, 160u);
	harness.memory.writeMappedU32LE(bmsx::IO_GX_CHARACTER_CELL_DATA, 0x2141u);
	const bmsx::GxCharacterPlaneOutput& output = harness.plane.readDeviceOutput();

	std::array<bmsx::u8, bmsx::GX_CHARACTER_PLANE_CELL_TEXTURE_BYTES> cellPixels{};
	std::array<bmsx::u8, bmsx::GX_CHARACTER_PLANE_GLYPH_TEXTURE_BYTES> glyphPixels{};
	std::array<bmsx::u8, bmsx::GX_CHARACTER_PLANE_PALETTE_TEXTURE_BYTES> palettePixels{};
	bmsx::writeGxCharacterPlaneCellTexture(output.cellBytes, cellPixels);
	bmsx::writeGxCharacterPlaneGlyphTexture(output.glyphBytes, glyphPixels);
	bmsx::writeGxCharacterPlanePaletteTexture(output.paletteBytes, palettePixels);
	require(cellPixels[0u] == 65u && cellPixels[1u] == 1u && cellPixels[2u] == 2u && cellPixels[3u] == 255u, "GX character cell texture encodes glyph and palette indexes");
	const size_t secondRowOffset = bmsx::GX_CHARACTER_PLANE_CELL_TEXTURE_WIDTH * 4u;
	require(cellPixels[secondRowOffset] == 65u && cellPixels[secondRowOffset + 1u] == 1u && cellPixels[secondRowOffset + 2u] == 2u, "GX character cell texture retains hardware row stride");
	const size_t glyphOffset = (24u * bmsx::GX_CHARACTER_PLANE_GLYPH_TEXTURE_WIDTH + 4u) * 4u;
	require(glyphPixels[glyphOffset] == 255u && glyphPixels[glyphOffset + 4u] == 0u, "GX character glyph texture expands packed mask bits");
	require(palettePixels[4u] == 255u && palettePixels[5u] == 0u && palettePixels[6u] == 0u && palettePixels[7u] == 255u, "GX character palette texture expands opaque red");
	require(palettePixels[8u] == 0u && palettePixels[9u] == 0u && palettePixels[10u] == 255u && palettePixels[11u] == 255u, "GX character palette texture expands opaque blue");

	std::array<bmsx::u32, 5u * 7u> framebuffer{};
	framebuffer.fill(0xff090a0bu);
	bmsx::SoftwareBackend backend(framebuffer.data(), 5, 7, 5 * static_cast<bmsx::i32>(sizeof(bmsx::u32)));
	bmsx::GxCharacterPlaneSoftwarePipeline pipeline;
	const bmsx::GxCharacterPlanePipelineState state{5, 7, &output};
	bmsx::renderGxCharacterPlaneSoftware(backend, pipeline, state);
	require(framebuffer[0u] == 0xffff0000u, "GX character software composition draws foreground mask bits");
	require(framebuffer[1u] == 0xff0000ffu, "GX character software composition draws background pixels");
	require(framebuffer[4u] == 0xff090a0bu, "GX character software composition preserves transparent clipped column");
	require(framebuffer[5u * 6u] == 0xffff0000u, "GX character software composition clips the final cell row");
}

} // namespace

int main() {
	testRawIndexedRegisterfiles();
	testDataReadNormalization();
	testStateRestore();
	testBackendResourcesAndSoftwareComposition();
	return 0;
}

File: tests/lua/gx_character_plane.test.ts
import assert from 'node:assert/strict';
import { test } from 'node:test';

import { readLE32 } from '../../machine/ts/common/endian';
import {
	IO_GX_CHARACTER_CELL_ADDRESS,
	IO_GX_CHARACTER_CELL_DATA,
	IO_GX_CHARACTER_CONTROL,
	IO_GX_CHARACTER_GLYPH_ADDRESS,
	IO_GX_CHARACTER_GLYPH_DATA,
	IO_GX_CHARACTER_PALETTE_ADDRESS,
	IO_GX_CHARACTER_PALETTE_DATA,
} from '../../machine/ts/machine/bus/io';
import {
	GX_CHARACTER_PLANE_CELL_WORDS,
	GX_CHARACTER_PLANE_CONTROL_ENABLE,
	GX_CHARACTER_PLANE_GLYPH_WORDS,
	GX_CHARACTER_PLANE_PALETTE_WORDS,
	GX_CHARACTER_PLANE_WORD_BYTES,
	GxCharacterPlane,
} from '../../machine/ts/machine/devices/gx/character_plane';
import { Memory } from '../../machine/ts/machine/memory/memory';
import {
	GxCharacterPlaneSoftwarePipeline,
	renderGxCharacterPlaneSoftware,
} from '../../machine/ts/render/backend/software/gx_character_plane';
import {
	GX_CHARACTER_PLANE_CELL_TEXTURE_BYTES,
	GX_CHARACTER_PLANE_CELL_TEXTURE_WIDTH,
	GX_CHARACTER_PLANE_GLYPH_TEXTURE_BYTES,
	GX_CHARACTER_PLANE_GLYPH_TEXTURE_WIDTH,
	GX_CHARACTER_PLANE_PALETTE_TEXTURE_BYTES,
	writeGxCharacterPlaneCellTexture,
	writeGxCharacterPlaneGlyphTexture,
	writeGxCharacterPlanePaletteTexture,
} from '../../machine/ts/render/gx/character_plane_resources';

function createCharacterPlane(): { memory: Memory; plane: GxCharacterPlane } {
	const memory = new Memory({
		systemRom: new Uint8Array(0),
		cartRom: new Uint8Array(0),
	});
	const plane = new GxCharacterPlane(memory);
	plane.reset();
	return { memory, plane };
}

test('GX character plane exposes raw indexed SRAM registerfiles', () => {
	const { memory, plane } = createCharacterPlane();
	const resetOutput = plane.readDeviceOutput();

	assert.equal(resetOutput.controlWord, 0);
	assert.equal(readLE32(resetOutput.paletteBytes, 0), 0);
	assert.equal(readLE32(resetOutput.glyphBytes, 0), 0);
	assert.equal(readLE32(resetOutput.cellBytes, 0), 0);

	memory.writeMappedU32LE(IO_GX_CHARACTER_CONTROL, (0x80000000 | GX_CHARACTER_PLANE_CONTROL_ENABLE) >>> 0);
	assert.equal(memory.readMappedU32LE(IO_GX_CHARACTER_CONTROL), 0x80000001);

	const paletteRevisionBeforeWrite = plane.readDeviceOutput().paletteRevision;
	const glyphRevisionBeforePaletteWrite = plane.readDeviceOutput().glyphRevision;
	const cellRevisionBeforePaletteWrite = plane.readDeviceOutput().cellRevision;
	memory.writeMappedU32LE(IO_GX_CHARACTER_PALETTE_ADDRESS, GX_CHARACTER_PLANE_PALETTE_WORDS - 1);
	memory.writeMappedU32LE(IO_GX_CHARACTER_PALETTE_DATA, 0x89abcdef);
	memory.writeMappedU32LE(IO_GX_CHARACTER_PALETTE_DATA, 0x01234567);
	assert.equal(memory.readMappedU32LE(IO_GX_CHARACTER_PALETTE_ADDRESS), 1);
	assert.equal(readLE32(resetOutput.paletteBytes, (GX_CHARACTER_PLANE_PALETTE_WORDS - 1) * GX_CHARACTER_PLANE_WORD_BYTES), 0x89abcdef);
	assert.equal(readLE32(resetOutput.paletteBytes, 0), 0x01234567);
	assert.notEqual(plane.readDeviceOutput().paletteRevision, paletteRevisionBeforeWrite);
	assert.equal(plane.readDeviceOutput().glyphRevision, glyphRevisionBeforePaletteWrite);
	assert.equal(plane.readDeviceOutput().cellRevision, cellRevisionBeforePaletteWrite);

	const paletteRevisionBeforeGlyphWrite = plane.readDeviceOutput().paletteRevision;
	const glyphRevisionBeforeWrite = plane.readDeviceOutput().glyphRevision;
	const cellRevisionBeforeGlyphWrite = plane.readDeviceOutput().cellRevision;
	memory.writeMappedU32LE(IO_GX_CHARACTER_GLYPH_ADDRESS, GX_CHARACTER_PLANE_GLYPH_WORDS - 1);
	memory.writeMappedU32LE(IO_GX_CHARACTER_GLYPH_DATA, 0xfedcba98);
	memory.writeMappedU32LE(IO_GX_CHARACTER_GLYPH_DATA, 0x76543210);
	assert.equal(memory.readMappedU32LE(IO_GX_CHARACTER_GLYPH_ADDRESS), 1);
	assert.equal(readLE32(resetOutput.glyphBytes, (GX_CHARACTER_PLANE_GLYPH_WORDS - 1) * GX_CHARACTER_PLANE_WORD_BYTES), 0xfedcba98);
	assert.equal(readLE32(resetOutput.glyphBytes, 0), 0x76543210);
	assert.equal(plane.readDeviceOutput().paletteRevision, paletteRevisionBeforeGlyphWrite);
	assert.notEqual(plane.readDeviceOutput().glyphRevision, glyphRevisionBeforeWrite);
	assert.equal(plane.readDeviceOutput().cellRevision, cellRevisionBeforeGlyphWrite);

	const paletteRevisionBeforeCellWrite = plane.readDeviceOutput().paletteRevision;
	const glyphRevisionBeforeCellWrite = plane.readDeviceOutput().glyphRevision;
	const cellRevisionBeforeWrite = plane.readDeviceOutput().cellRevision;
	memory.writeMappedU32LE(IO_GX_CHARACTER_CELL_ADDRESS, GX_CHARACTER_PLANE_CELL_WORDS - 1);
	memory.writeMappedU32LE(IO_GX_CHARACTER_CELL_DATA, 0xdeadbeef);
	memory.writeMappedU32LE(IO_GX_CHARACTER_CELL_DATA, 0x13579bdf);
	assert.equal(memory.readMappedU32LE(IO_GX_CHARACTER_CELL_ADDRESS), 1);
	assert.equal(readLE32(resetOutput.cellBytes, (GX_CHARACTER_PLANE_CELL_WORDS - 1) * GX_CHARACTER_PLANE_WORD_BYTES), 0xdeadbeef);
	assert.equal(readLE32(resetOutput.cellBytes, 0), 0x13579bdf);
	assert.equal(plane.readDeviceOutput().paletteRevision, paletteRevisionBeforeCellWrite);
	assert.equal(plane.readDeviceOutput().glyphRevision, glyphRevisionBeforeCellWrite);
	assert.notEqual(plane.readDeviceOutput().cellRevision, cellRevisionBeforeWrite);

	const output = plane.readDeviceOutput();
	assert.equal(output, resetOutput);
	const paletteRevisionBeforeReset = output.paletteRevision;
	const glyphRevisionBeforeReset = output.glyphRevision;
	const cellRevisionBeforeReset = output.cellRevision;
	plane.reset();
	assert.equal(plane.readDeviceOutput().controlWord, 0);
	assert.equal(readLE32(output.paletteBytes, 0), 0);
	assert.equal(readLE32(output.glyphBytes, 0), 0);
	assert.equal(readLE32(output.cellBytes, 0), 0);
	assert.notEqual(output.paletteRevision, paletteRevisionBeforeReset);
	assert.notEqual(output.glyphRevision, glyphRevisionBeforeReset);
	assert.notEqual(output.cellRevision, cellRevisionBeforeReset);
});

test('GX character plane data reads normalize and advance raw address latches', () => {
	const { memory } = createCharacterPlane();

	memory.writeMappedU32LE(IO_GX_CHARACTER_PALETTE_ADDRESS, 0xffffffff);
	memory.writeMappedU32LE(IO_GX_CHARACTER_PALETTE_DATA, 0xa1b2c3d4);
	memory.writeMappedU32LE(IO_GX_CHARACTER_PALETTE_ADDRESS, 0xffffffff);
	assert.equal(memory.readMappedU32LE(IO_GX_CHARACTER_PALETTE_DATA), 0xa1b2c3d4);
	assert.equal(memory.readMappedU32LE(IO_GX_CHARACTER_PALETTE_ADDRESS), 0);

	memory.writeMappedU32LE(IO_GX_CHARACTER_CELL_ADDRESS, 0xffffffff);
	memory.writeMappedU32LE(IO_GX_CHARACTER_CELL_DATA, 0x2468ace0);
	memory.writeMappedU32LE(IO_GX_CHARACTER_CELL_ADDRESS, 0xffffffff);
	assert.equal(memory.readMappedU32LE(IO_GX_CHARACTER_CELL_DATA), 0x2468ace0);
	assert.equal(memory.readMappedU32LE(IO_GX_CHARACTER_CELL_ADDRESS), (0xffffffff % GX_CHARACTER_PLANE_CELL_WORDS) + 1);
});

test('GX character plane state restores SRAM and address latches in place', () => {
	const { memory, plane } = createCharacterPlane();

	memory.writeMappedU32LE(IO_GX_CHARACTER_CONTROL, 0xf0000001);
	memory.writeMappedU32LE(IO_GX_CHARACTER_PALETTE_ADDRESS, 3);
	memory.writeMappedU32LE(IO_GX_CHARACTER_PALETTE_DATA, 0x80007c00);
	memory.writeMappedU32LE(IO_GX_CHARACTER_GLYPH_ADDRESS, 65);
	memory.writeMappedU32LE(IO_GX_CHARACTER_GLYPH_DATA, 0x00f99f90);
	memory.writeMappedU32LE(IO_GX_CHARACTER_CELL_ADDRESS, 321);
	memory.writeMappedU32LE(IO_GX_CHARACTER_CELL_DATA, 0xabcd1241);
	const state = plane.captureState();
	const output = plane.readDeviceOutput();
	const paletteBytes = output.paletteBytes;
	const glyphBytes = output.glyphBytes;
	const cellBytes = output.cellBytes;

	memory.writeMappedU32LE(IO_GX_CHARACTER_CONTROL, 0);
	memory.writeMappedU32LE(IO_GX_CHARACTER_PALETTE_ADDRESS, 3);
	memory.writeMappedU32LE(IO_GX_CHARACTER_PALETTE_DATA, 0);
	memory.writeMappedU32LE(IO_GX_CHARACTER_GLYPH_ADDRESS, 65);
	memory.writeMappedU32LE(IO_GX_CHARACTER_GLYPH_DATA, 0);
	memory.writeMappedU32LE(IO_GX_CHARACTER_CELL_ADDRESS, 321);
	memory.writeMappedU32LE(IO_GX_CHARACTER_CELL_DATA, 0);
	const changedRevision = plane.readDeviceOutput().cellRevision;

	plane.restoreState(state);
	const restored = plane.readDeviceOutput();
	assert.equal(restored.paletteBytes, paletteBytes);
	assert.equal(restored.glyphBytes, glyphBytes);
	assert.equal(restored.cellBytes, cellBytes);
	assert.equal(restored.controlWord, 0xf0000001);
	assert.equal(memory.readMappedU32LE(IO_GX_CHARACTER_PALETTE_ADDRESS), 4);
	assert.equal(memory.readMappedU32LE(IO_GX_CHARACTER_GLYPH_ADDRESS), 66);
	assert.equal(memory.readMappedU32LE(IO_GX_CHARACTER_CELL_ADDRESS), 322);
	assert.equal(readLE32(restored.paletteBytes, 3 * GX_CHARACTER_PLANE_WORD_BYTES), 0x80007c00);
	assert.equal(readLE32(restored.glyphBytes, 65 * GX_CHARACTER_PLANE_WORD_BYTES), 0x00f99f90);
	assert.equal(readLE32(restored.cellBytes, 321 * GX_CHARACTER_PLANE_WORD_BYTES), 0xabcd1241);
	assert.notEqual(restored.cellRevision, changedRevision);
});

test('GX character plane backend resources and software composition preserve transparency and clipping', () => {
	const { memory, plane } = createCharacterPlane();
	memory.writeMappedU32LE(IO_GX_CHARACTER_PALETTE_ADDRESS, 1);
	memory.writeMappedU32LE(IO_GX_CHARACTER_PALETTE_DATA, 0x801f);
	memory.writeMappedU32LE(IO_GX_CHARACTER_PALETTE_DATA, 0xfc00);
	memory.writeMappedU32LE(IO_GX_CHARACTER_GLYPH_ADDRESS, 65);
	memory.writeMappedU32LE(IO_GX_CHARACTER_GLYPH_DATA, 1);
	memory.writeMappedU32LE(IO_GX_CHARACTER_CELL_ADDRESS, 0);
	memory.writeMappedU32LE(IO_GX_CHARACTER_CELL_DATA, 0x2141);
	memory.writeMappedU32LE(IO_GX_CHARACTER_CELL_ADDRESS, 160);
	memory.writeMappedU32LE(IO_GX_CHARACTER_CELL_DATA, 0x2141);
	const output = plane.readDeviceOutput();

	const cellPixels = new Uint8Array(GX_CHARACTER_PLANE_CELL_TEXTURE_BYTES);
	const glyphPixels = new Uint8Array(GX_CHARACTER_PLANE_GLYPH_TEXTURE_BYTES);
	const palettePixels = new Uint8Array(GX_CHARACTER_PLANE_PALETTE_TEXTURE_BYTES);
	writeGxCharacterPlaneCellTexture(output.cellBytes, cellPixels);
	writeGxCharacterPlaneGlyphTexture(output.glyphBytes, glyphPixels);
	writeGxCharacterPlanePaletteTexture(output.paletteBytes, palettePixels);
	assert.deepEqual(Array.from(cellPixels.subarray(0, 4)), [65, 1, 2, 255]);
	assert.deepEqual(Array.from(cellPixels.subarray(GX_CHARACTER_PLANE_CELL_TEXTURE_WIDTH * 4, GX_CHARACTER_PLANE_CELL_TEXTURE_WIDTH * 4 + 4)), [65, 1, 2, 255]);
	const glyphOffset = (24 * GX_CHARACTER_PLANE_GLYPH_TEXTURE_WIDTH + 4) * 4;
	assert.deepEqual(Array.from(glyphPixels.subarray(glyphOffset, glyphOffset + 8)), [255, 0, 0, 255, 0, 0, 0, 255]);
	assert.deepEqual(Array.from(palettePixels.subarray(4, 12)), [255, 0, 0, 255, 0, 0, 255, 255]);

	const target = new Uint8Array(5 * 7 * 4);
	for (let offset = 0; offset < target.byteLength; offset += 4) {
		target[offset] = 9;
		target[offset + 1] = 10;
		target[offset + 2] = 11;
		target[offset + 3] = 255;
	}
	renderGxCharacterPlaneSoftware(new GxCharacterPlaneSoftwarePipeline(), { width: 5, height: 7, output }, target);
	assert.deepEqual(Array.from(target.subarray(0, 8)), [255, 0, 0, 255, 0, 0, 255, 255]);
	assert.deepEqual(Array.from(target.subarray(16, 20)), [9, 10, 11, 255]);
	assert.deepEqual(Array.from(target.subarray(5 * 6 * 4, 5 * 6 * 4 + 4)), [255, 0, 0, 255]);
});
